altair/utils/core.py

```diff
@@ -53,6 +52,11 @@ def __dataframe__(
     ) -> DfiDataFrame: ...
 
 
+def _is_pandas_dataframe(obj: Any) -> bool:
```

Comment:
Maybe I'm missing something but couldn't we call the pandas import function you created in here and if it raises an importerror, we know it's not a pandas dataframe anyway.

Reply:
It's pragmatic, I admit. But that would be an unnecessary import of pandas if it is available in the environment, but if the data object is something else.
I wish we could sniff the type without importing modules first.

Reply:
Here's the optional import logic I added to plotly.py a while back: https://github.com/plotly/plotly.py/blob/master/packages/python/plotly/_plotly_utils/optional_imports.py if `should_load` is `False` then it won't perform the import even if the library is installed. This was used with `isinstance` checks, because if `pandas` hasn't been loaded yet, you know the object you're dealing with isn't a pandas DataFrame, even if pandas is installed.

Reply:
A trick I learned from scikit-learn is to check if `pandas` is in `sys.modules` before doing the isinstance check, something like
```
if (pd := sys.modules.get('pandas')) is not None and isinstance(df, pd.DataFrame):
    ...
```
If pandas was never imported, then `df` is definitely not pandas

Reply:
I just see your response here @MarcoGorelli! I also made this observation recently, see the comment I just added https://github.com/vega/altair/pull/3384#issuecomment-2191589724...